pyproject.toml

```diff
@@ -12,42 +12,68 @@ authors = [
 requires-python = "==3.12.*"
 license = { text = "BSD-3-Clause" }
 dependencies = [
-  "appdirs>=1.4",
-  "azure-core>=1.26",
-  "azure-identity>=1.16.1",
-  "azure-keyvault-certificates>=4.6",
-  "azure-keyvault-keys>=4.6",
-  "azure-keyvault-secrets>=4.6",
-  "azure-mgmt-automation>=1.0",
-  "azure-mgmt-compute>=30.3",
-  "azure-mgmt-containerinstance>=10.1",
-  "azure-mgmt-dns>=8.0",
-  "azure-mgmt-keyvault>=10.3",
-  "azure-mgmt-msi>=7.0",
-  "azure-mgmt-network>=25.0",
-  "azure-mgmt-rdbms>=10.1",
-  "azure-mgmt-resource>=23.0",
-  "azure-mgmt-storage>=21.1",
-  "azure-storage-blob>=12.15",
-  "azure-storage-file-datalake>=12.10",
-  "azure-storage-file-share>=12.10",
-  "chevron>=0.14",
-  "cryptography>=42.0",
-  "fqdn>=1.5",
-  "msal>=1.21",
-  "psycopg>=3.1",
-  "pulumi>=3.80",
-  "pulumi-azure-native>=2.49.1",
-  "pulumi-random>=4.14",
-  "pulumi-tls>=5.0.3",
-  "pydantic>=2.4",
-  "pyjwt>=2.8",
-  "pytz>=2023.3",
-  "PyYAML>=6.0",
-  "rich>=13.4",
-  "simple-acme-dns>=3.0",
-  "typer>=0.9",
-  "websocket-client>=1.5",
+  "appdirs==1.4.4",
+  "azure-core==1.30.2",
```

Comment:
I still don't see the advantage of pinning dependencies both here and in the `requirements.txt` files. Won't this just make it more difficult to resolve a full set of dependencies?

You can't tell which are these are strict `==` and which at actually `>=` some version.

Reply:
- the point of pinning requirements here is to have a single place where the direct requirements of this project are pinned
- the requirements files are just a way to ensure that the indirect requirements are also pinned (using `pip-compile`)

Ideally there would be a way to pin the indirect dependencies in `pyproject.toml` too, but this method is the one recommended by the `hatch` developers (see [here](https://github.com/pypa/hatch/issues/749))

Reply:
It's a bit of a conceptual shift away from "which versions of our dependencies will work" to "which versions of our dependencies to we want users to install". It wouldn't make sense if we were expecting people to run `from package_name import whatever` but it does for a package that we want to be installed in its own, dedicated venv.

Reply:
If there were a way to get `pip install .` and `hatch env create` to have _exactly_ that same dependencies (without putting a full pinned list in `pyproject.toml`), that would be good. It feels like that might not be an option though.

Reply:
We sort of want the opposite of what we are doing here (like the pip-tools docs recommends) a `requirements.in` which we use to generate the dependencies in `pyproject.toml`.

We _could_ presumably write that ourselves. Using `pip-compile` as a library to generate the pinned dependencies, then a toml library to write that list to `pyproject.toml`.